napari/layers/graph/graph.py

```diff
@@ -0,0 +1,508 @@
+        self.data.add_nodes(indices=indices, coords=coords)
+        self._data_changed(prev_size)
+
+    def remove_selected(self) -> None:
+        """Removes selected points if any."""
+        if len(self.selected_data):
+            self._remove_nodes(list(self.selected_data), is_buffer_domain=True)
+            self.selected_data = cast(Selection[int], set())
+
+    def remove(self, indices: ArrayLike) -> None:
+        """Remove nodes given indices."""
+        self._remove_nodes(indices, is_buffer_domain=False)
+
+    def _remove_nodes(
+        self,
+        indices: ArrayLike,
+        is_buffer_domain: bool,
+    ) -> None:
+        """
+        Parameters
+        ----------
+        indices : ArrayLike
+            List of node indices to remove.
+        is_buffer_domain : bool
+            Indicates if node indices are on world or buffer domain.
+        """
+        indices = np.atleast_1d(indices)
+        if indices.ndim > 1:
+            raise ValueError(
+                trans._(
+                    "Indices for removal must be 1-dim. Found {ndim}",
+                    ndim=indices.ndim,
+                )
+            )
+
+        prev_size = self.data.n_allocated_nodes
+
+        # it got error missing __iter__ attribute, but we guarantee by np.atleast_1d call
+        for idx in indices:  # type: ignore[union-attr]
+            self.data.remove_node(idx, is_buffer_domain)
+
+        self._data_changed(prev_size)
+
+    def _move_points(
+        self, ixgrid: Tuple[np.ndarray, np.ndarray], shift: np.ndarray
+    ) -> None:
+        """Move points along a set a coordinates given a shift.
+
+        Parameters
+        ----------
+        ixgrid : Tuple[np.ndarray, np.ndarray]
+            Crossproduct indexing grid of node indices and dimensions, see `np.ix_`
+        shift : np.ndarray
+            Selected coordinates shift
+        """
+        self.data.coords_buffer[ixgrid] = (
+            self.data.coords_buffer[ixgrid] + shift
+        )
+
+    def _update_props_and_style(self, data_size: int, prev_size: int) -> None:
+        # Add/remove property and style values based on the number of new points.
+        with self.events.blocker_all(), self._border.events.blocker_all(), self._face.events.blocker_all():
+            self._feature_table.resize(data_size)
+            self.text.apply(self.features)
+            if data_size < prev_size:
+                # If there are now fewer points, remove the size and colors of the
+                # extra ones
+                if len(self._border.colors) > data_size:
+                    self._border._remove(
+                        np.arange(data_size, len(self._border.colors))
+                    )
+                if len(self._face.colors) > data_size:
+                    self._face._remove(
+                        np.arange(data_size, len(self._face.colors))
+                    )
+                self._shown = self._shown[:data_size]
```

Comment:
This is a little odd because we're just trimming properties off the end rather than removing the property of the actual point. Leads to some strange behaviour when removing and then adding points.

Reply:
Agreed it's strange, but it [is consistent with `Points`](https://github.com/napari/napari/blob/787fc5f450aeb1d7ea23a7915f096fd5185f40b2/napari/layers/points/points.py#L593).

Similar to [the discussion on the data setter and refresh](https://github.com/napari/napari/pull/5861/files/ea2ac8c9364855cfff62a5692aab7f214ddd95bc#r1336217787), do we want to hold the `Graph` layer to a higher standard than the other layers and fix those underlying problems first. Or should we get it in with known issues and potential workarounds, and use that motivate fixes for those underlying problems?